localtileserver/custom/widgets.py

```diff
@@ -0,0 +1,76 @@
+import concurrent.futures
+import os
+import pathlib
+import uuid
+
+import anywidget
+import traitlets
+
+from localtileserver.client import TileClient
+
+THIS_DIR = pathlib.Path(__file__).parent.absolute()
+THREAD_EXECUTOR = concurrent.futures.ThreadPoolExecutor(max_workers=os.cpu_count())
+
+
+def handle_anywidget_command(widget, msg: str | list | dict, buffers: list[bytes]) -> None:
+    if not isinstance(msg, dict) or msg.get("kind") != "anywidget-command":
+        return
+
+    def target():
+        response, buffers = widget.get_tile(msg["msg"])
+        widget.send(
+            {
+                "id": msg["id"],
+                "kind": "anywidget-command-response",
+                "response": response,
+            },
+            buffers,
+        )
+
+    # t = threading.Thread(target=target)
+    # t.start()
+    THREAD_EXECUTOR.submit(target)
+
+
+class TileLayerWidget(anywidget.AnyWidget):
+    """Leaflet Tile Layer Widget.
+
+    Args:
+        anywidget (_type_): _description_
+    """
+
+    _esm = THIS_DIR / "lts_widget.js"
+    _css = THIS_DIR / "lts_widget.css"
+
+    bounds = traitlets.List([0, 0, 0, 0]).tag(sync=True, o=True)
+    center = traitlets.List([0, 0]).tag(sync=True, o=True)
+    identifier = traitlets.Unicode("").tag(sync=True, o=True)
+    default_zoom = traitlets.Int(0).tag(sync=True, o=True)
+
+    def __init__(self, client: TileClient, **kwargs):
+        super().__init__()
+        self.client = client  # TODO: weakref?
+        self.on_msg(handle_anywidget_command)
+        self._display_kwargs = kwargs
+        self.bounds = self.client.bounds()
+        self.identifier = uuid.uuid4().hex
+        self.center = self.client.center()
+        self.default_zoom = self.client.default_zoom
+
+    def get_tile(self, msg):
+        try:
+            x = int(msg["x"])
+            y = int(msg["y"])
+            z = int(msg["z"])
+        except (KeyError, ValueError) as e:
+            return f"failed: {e}", []
+
+        try:
+            # NOTE: when calling from a thread executor, we need a different dataset handle for each thread
+            # finding that the kernel crashes... is rio-tiler or rasterio not thread safe?
```

Comment:
No, definitely not thread safe by default. https://github.com/gjoseph92/stackstac/blob/main/stackstac/rio_reader.py Is a really cool implementation of threading for some specific rasterio drivers (e.g. COG)